File: src/main/java/systems/conduit/main/mixins/level/LevelMixin.java
package systems.conduit.main.mixins.level;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.DifficultyInstance;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.GameRules;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.biome.BiomeManager;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.border.WorldBorder;
import net.minecraft.world.level.chunk.LevelChunk;
import net.minecraft.world.level.dimension.DimensionType;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.storage.LevelData;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import systems.conduit.main.core.api.mixins.Biome;

import java.util.Optional;

/**
 * The exposed level API.
 *
 * @author Innectic
 * @since 11/28/2019
 */
@Mixin(value = Level.class, remap = false)
public abstract class LevelMixin implements systems.conduit.main.core.api.mixins.Level {

    @Shadow public abstract BiomeManager shadow$getBiomeManager();

    @Override
    public Biome getBiome(BlockPos pos) {
        return (Biome) (Object) this.shadow$getBiomeManager().getBiome(pos);
    }

    @Shadow public abstract LevelChunk getChunkAt(BlockPos pos);
    @Shadow public abstract LevelChunk getChunk(int x, int z);
    @Shadow public abstract boolean setBlock(BlockPos pos, BlockState state, int flag);
    @Shadow public abstract boolean removeBlock(BlockPos pos, boolean flag);
    @Shadow public abstract boolean setBlockAndUpdate(BlockPos pos, BlockState state);
    @Shadow public abstract BlockState getBlockState(BlockPos pos);
    @Shadow public abstract FluidState getFluidState(BlockPos pos);
    @Shadow public abstract boolean isDay();
    @Shadow public abstract void playSound(Player player, BlockPos pos, SoundEvent event, SoundSource source, float pitch, float volume);
    @Shadow public abstract BlockEntity shadow$getBlockEntity(BlockPos pos);
    @Shadow public abstract void setBlockEntity(BlockEntity entity);
    @Shadow public abstract void removeBlockEntity(BlockPos pos);
    @Shadow public abstract boolean isLoaded(BlockPos pos);
    @Shadow public abstract int getSeaLevel();
    @Shadow public abstract DimensionType dimensionType();
    @Shadow public abstract boolean hasSignal(BlockPos pos, Direction direction);
    @Shadow public abstract int getSignal(BlockPos pos, Direction direction);
    @Shadow public abstract boolean hasNeighborSignal(BlockPos pos);
    @Shadow public abstract int getBestNeighborSignal(BlockPos pos);
//    @Shadow public abstract long getSeed();
    @Shadow public abstract long getGameTime();
    @Shadow public abstract long getDayTime();
//    @Shadow public abstract void setDayTime(long time);
//    @Shadow public abstract BlockPos getSharedSpawnPos();
    @Shadow public abstract LevelData getLevelData();
    @Shadow public abstract GameRules getGameRules();
    @Shadow public abstract boolean isThundering();
    @Shadow public abstract boolean isRaining();
    @Shadow public abstract boolean isRainingAt(BlockPos pos);
    @Shadow public abstract boolean isHumidAt(BlockPos pos);
//    @Shadow public abstract int getHeight();
    @Shadow public abstract DifficultyInstance getCurrentDifficultyAt(BlockPos pos);
    @Shadow public abstract int getSkyDarken();
    @Shadow public abstract void setSkyFlashTime(int time);
    @Shadow public abstract WorldBorder getWorldBorder();

    public Optional<BlockEntity> getBlockEntity(BlockPos pos) {
        return Optional.ofNullable(shadow$getBlockEntity(pos));
    }

}

File: src/main/kotlin/systems/conduit/core/mixins/api/LevelMixin.kt
package systems.conduit.core.mixins.api

import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.sounds.SoundEvent
import net.minecraft.sounds.SoundSource
import net.minecraft.world.DifficultyInstance
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.GameRules
import net.minecraft.world.level.LevelType
import net.minecraft.world.level.LightLayer
import net.minecraft.world.level.biome.Biome
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.border.WorldBorder
import net.minecraft.world.level.chunk.ChunkSource
import net.minecraft.world.level.chunk.LevelChunk
import net.minecraft.world.level.dimension.Dimension
import net.minecraft.world.level.material.FluidState
import net.minecraft.world.level.storage.LevelData
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import systems.conduit.api.Level

/**
 * The exposed level API.
 *
 * @author Innectic
 * @since 11/28/2019
 */
@Mixin(value = [net.minecraft.world.level.Level::class], remap = false)
abstract class LevelMixin: Level {

    @Shadow
    abstract fun `shadow$getBiome`(pos: BlockPos?): Biome?

    @Shadow
    abstract override fun getTopBlockState(pos: BlockPos?): BlockState?

    @Shadow
    abstract override fun getChunkAt(pos: BlockPos?): LevelChunk?

    @Shadow
    abstract override fun getChunk(x: Int, z: Int): LevelChunk?

    @Shadow
    abstract override fun setBlock(pos: BlockPos?, state: BlockState?, flag: Int): Boolean

    @Shadow
    abstract override fun removeBlock(pos: BlockPos?, flag: Boolean): Boolean

    @Shadow
    abstract override fun destroyBlock(pos: BlockPos?, dropItems: Boolean): Boolean

    @Shadow
    abstract override fun setBlockAndUpdate(pos: BlockPos?, state: BlockState?): Boolean

    @Shadow
    abstract override fun getRawBrightness(pos: BlockPos?, mask: Int): Int

    @Shadow
    abstract override fun getBrightness(layer: LightLayer?, pos: BlockPos?): Int

    @Shadow
    abstract override fun getBlockState(pos: BlockPos?): BlockState?

    @Shadow
    abstract override fun getFluidState(pos: BlockPos?): FluidState?

    @Shadow
    abstract override fun isDay(): Boolean

    @Shadow
    abstract override fun playSound(player: Player?, pos: BlockPos?, event: SoundEvent?, source: SoundSource?, pitch: Float, volume: Float)

    @Shadow
    abstract fun `shadow$getBlockEntity`(pos: BlockPos?): BlockEntity?

    @Shadow
    abstract override fun setBlockEntity(pos: BlockPos?, entity: BlockEntity?)

    @Shadow
    abstract override fun removeBlockEntity(pos: BlockPos?)

    @Shadow
    abstract override fun isLoaded(pos: BlockPos?): Boolean

    @Shadow
    abstract override fun getSeaLevel(): Int

    @Shadow
    abstract override fun getGeneratorType(): LevelType?

    @Shadow
    abstract override fun hasSignal(pos: BlockPos?, direction: Direction?): Boolean

    @Shadow
    abstract override fun getSignal(pos: BlockPos?, direction: Direction?): Int

    @Shadow
    abstract override fun hasNeighborSignal(pos: BlockPos?): Boolean

    @Shadow
    abstract override fun getBestNeighborSignal(pos: BlockPos?): Int

    @Shadow
    abstract override fun setGameTime(time: Long)

    @Shadow
    abstract override fun getSeed(): Long

    @Shadow
    abstract override fun getGameTime(): Long

    @Shadow
    abstract override fun getDayTime(): Long

    @Shadow
    abstract override fun setDayTime(time: Long)

    @Shadow
    abstract override fun getSharedSpawnPos(): BlockPos?

    @Shadow
    abstract override fun setSpawnPos(pos: BlockPos?)

    @Shadow
    abstract override fun getChunkSource(): ChunkSource?

    @Shadow
    abstract override fun getLevelData(): LevelData?

    @Shadow
    abstract override fun getGameRules(): GameRules?

    @Shadow
    abstract override fun isThundering(): Boolean

    @Shadow
    abstract override fun isRaining(): Boolean

    @Shadow
    abstract override fun isRainingAt(pos: BlockPos?): Boolean

    @Shadow
    abstract override fun isHumidAt(pos: BlockPos?): Boolean

    @Shadow
    abstract override fun getHeight(): Int

    @Shadow
    abstract override fun getCurrentDifficultyAt(pos: BlockPos?): DifficultyInstance?

    @Shadow
    abstract override fun getSkyDarken(): Int

    @Shadow
    abstract override fun setSkyFlashTime(time: Int)

    @Shadow
    abstract override fun getWorldBorder(): WorldBorder?

    @Shadow
    abstract override fun getDimension(): Dimension?
    override fun getBiome(pos: BlockPos?): systems.conduit.api.Biome? {
        return `shadow$getBiome`(pos) as systems.conduit.api.Biome?
    }

    override fun getBlockEntity(pos: BlockPos?): BlockEntity? {
        return `shadow$getBlockEntity`(pos)
    }
}
